plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -282,15 +356,78 @@ fu_rts54hub_rtd21xx_ensure_version_unlocked (FuRts54hubRtd21xxDevice *self,
 	return TRUE;
 }
 
+static gboolean
+fu_rts54hub_rtd21xx_ensure_version_background (FuRts54hubRtd21xxDevice *self,
+					     GError **error)
+{
+	guint8 buf_rep[7] = { 0x00 };
+	guint8 buf_req[] = { ISP_BG_CMD_GET_FW_INFO };
+	guint8 buf[] = { ISP_BG_CMD_FW_UPDATE_EXIT };
+	g_autofree gchar *version = NULL;
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   buf_req, sizeof(buf_req),
+						   error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	g_usleep (300000);
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (self,
+						  UC_ISP_SLAVE_ADDR,
+						  0x00,
+						  buf_rep, sizeof(buf_rep),
+						  error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+	/* set version */
+	version = g_strdup_printf ("%u.%u", buf_rep[1], buf_rep[2]);
+	fu_device_set_version (FU_DEVICE (self), version);
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   buf, sizeof(buf),
+						   error)) {
+		g_prefix_error (error, "failed to ISP_CMD_FW_UPDATE_RESET: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_ensure_version_unlocked (FuRts54hubRtd21xxDevice *self,
+					     GError **error)
+{
+
+	if (self->fwupdate_mode == FW_UPDATE_BACKGROUND_MODE)
+		return fu_rts54hub_rtd21xx_ensure_version_background (self, error);
+	else
+		return fu_rts54hub_rtd21xx_ensure_version_foreground (self, error);
+	return TRUE;
+}
+
 static gboolean
 fu_rts54hub_rtd21xx_device_detach_raw (FuRts54hubRtd21xxDevice *self, GError **error)
 {
 	guint8 buf = 0x03;
+
+	if (self->fwupdate_mode == FW_UPDATE_BACKGROUND_MODE)
+		buf = 0x02;
+	else
+		buf = 0x03;
```

Comment:
you seem to set this default in the initialization.  I would suggest pick one or the other.

Reply:
OK ~~